core/cli/node/client.hpp

```diff
@@ -0,0 +1,854 @@
+               "Getting IPLD id of data associated with provided address...");
+    const Actor actor =
+        cliTry(api->StateGetActor(kVerifiedRegistryAddress, TipsetKey()),
+               "Getting VerifierActor");
+    const auto ipfs = std::make_shared<ApiIpfsDatastore>(api);
+    const NetworkVersion version = cliTry(api->StateNetworkVersion(TipsetKey()),
+                                          "Getting Chain Version...");
+    ipfs->actor_version = actorVersion(version);
+    const VerifiedRegistryActorStatePtr state =
+        cliTry(getCbor<VerifiedRegistryActorStatePtr>(ipfs, actor.head));
+    auto res = cliTry(cliTry(state->getVerifierDataCap(vid)),
+                      "Client {} isn't in notary tables",
+                      vaddr);
+    return res;
+  }
+
+  struct Node_client_retrieve {
+    struct Args {
+      CLI_OPTIONAL("from", "address to send transactions from", Address) from;
+      CLI_OPTIONAL("provider",
+                   "provider to use for retrieval, if not present it'll use "
+                   "local discovery",
+                   Address)
+      provider;
+      CLI_OPTIONAL("pieceCid",
+                   "require data to be retrieved from a specific Piece CID",
+                   CID)
+      piece_cid;
+      CLI_DEFAULT(
+          "maxPrice",
+          fmt::format(
+              "maximum price the client is willing to consider (default: {}"
+              " FIL)",
+              kDefaultMaxRetrievePrice.atto())
+              .c_str(),
+          AttoFil,
+          {kDefaultMaxRetrievePrice})
+      max_price;
+      CLI_OPTIONAL("data-selector",
+                   "(NOT SUPPORTED) IPLD datamodel text-path selector, or IPLD "
+                   "json selector",
+                   std::string)
+      data_selector;
+      CLI_BOOL("car", "Export to a car file instead of a regular file") car;
+      CLI_BOOL("allow-local", "allow use local imports") allow_local;
+      CLI_BOOL("car-export-merkle-proof",
+               "(requires --data-selector and --car) Export data-selector "
+               "merkle proof")
+      car_export_merkle_proof;
+
+      CLI_OPTS() {
+        Opts opts;
+        from(opts);
+        provider(opts);
+        piece_cid(opts);
+        allow_local(opts);
+        car(opts);
+        max_price(opts);
+        data_selector(opts);
+        car_export_merkle_proof(opts);
+        return opts;
+      }
+    };
+
+    CLI_RUN() {
+      const Node::Api api{argm};
+
+      const CID data_cid{cliArgv<CID>(argv, 0, "dataCid")};
+      const std::string path{cliArgv<std::string>(argv, 1, "path")};
+
+      const boost::optional<CID> piece_cid = args.piece_cid.v;
+      RetrievalOrder order{
+          .client =
+              (args.from ? *args.from
+                         : cliTry(api->WalletDefaultAddress(),
+                                  "Getting address of default wallet...")),
+          .miner = (args.provider ? *args.provider : Address{})};
```

Comment:
`.miner` may be uninitialized. Please move filling `order` and `.miner` from `offer` after getting `offer`
```suggestion
```

Reply:
You mean that if args.provider is false then .miner not needed in fill empty address like Address{} ?

Reply:
Lotus always initializes `.miner` from `offset`, but this code may leave it unitialized

Reply:
Fixed